### PR TITLE
Make Omni cluster templates fail loudly on missing/out-of-range CIDRs

### LINKED ISSUE #1086: Make the Omni cluster template base fail loudly when per-cluster CIDRs are missing
> [!NOTE]
> **TL;DR**: The reusable Omni cluster template base should refuse to validate (or scream in its docs) when copied without the per-cluster pod/service CIDRs and clusterDNS, so ADR-014-violating Talos defaults can't sneak in silently.

## Context

`catalog/omni/clustertemplates/base.yaml` (introduced on the `i1038-add-omni-cluster-template` branch, commit `a2a94a80e`) is a reusable reference for Talos-on-Proxmox-VE clusters. It deliberately omits `cluster.network.podSubnets`, `cluster.network.serviceSubnets`, and `machine.kubelet.clusterDNS` because those are per-cluster values — its header lists them as "override at minimum".

The problem: they are **additions**, not value changes. A copyist who forgets them gets the Talos machine-config defaults (`10.244.0.0/16` pods, `10.96.0.0/12` services, `10.96.0.10` DNS), which directly violate ADR-014's `172.16.0.0/12`-only allocation rule. `omnictl cluster template validate` passes either way — the omission is **silent**. The lungmen template (`projects/lungmen.akn/src/infrastructure/omni/lungmen.clustertemplate.yaml`) correctly adds all three, so this is a base-robustness gap, not a live defect.

This was flagged as a **minor** finding in the #1038 code review (Parse-Don't-Validate / Fail-Fast laws: the illegal state is currently representable and silent).

## Proposal

Two approaches to weigh (or combine):

| Option | How | Trade-off |
| --- | --- | --- |
| (a) Sentinel placeholders | Add `podSubnets: ["MUST-OVERRIDE/19"]` etc. so `omnictl validate` fails until overridden | Maximal fail-loud, but the reference base itself no longer validates (the documented `validate` command would then fail on the base) |
| (b) Explicit defaults + loud comment | Add the three fields with Talos default values plus a `# MUST OVERRIDE — Talos defaults violate ADR-014` comment, and strengthen the header checklist | Base stays self-validating; copyist still has to notice the comment (softer than (a)) |

A third option worth considering: a small CI-time lint (OPA/Rego in `catalog/opa/`, or a `dist:render`-adjacent check) that fails any per-cluster template missing the CIDR fields — keeping the base a pure reference while enforcing correctness on real instances.

## Acceptance criteria

- [ ] A copied-and-forgotten cluster template (missing podSubnets/serviceSubnets/clusterDNS) is no longer silently accepted — either validate fails loudly, or the omission is caught by lint/docs review.
- [ ] The reference base still documents its own validate/apply workflow coherently (no broken documented command).
- [ ] ADR-014's CIDR rule is mechanically enforced for per-cluster templates, not just documented.

## Notes

- Surfaced during the #1038 review of `catalog/omni/clustertemplates/base.yaml` (commit `a2a94a80e`).
- ADR-014: `docs/decisions/014-network-topology.md`.
- Worked correct example: `projects/lungmen.akn/src/infrastructure/omni/lungmen.clustertemplate.yaml` (podSubnets `172.30.0.0/19`, serviceSubnets `172.31.0.0/19`, clusterDNS `172.31.0.10`).

---

<sub>AI-assisted with zai:glm-5.2 under human supervision</sub>

### PR DESCRIPTION
**Pull request by xunleii**
## Summary

`catalog/omni/clustertemplates/base.yaml` (the reusable Talos-on-Proxmox-VE reference template) deliberately omits
`cluster.network.podSubnets` because it's per-cluster — but the omission was silent: a copyist who forgot to override
it got Talos's built-in pod CIDR (`10.244.0.0/16`), which directly violates ADR-014's `172.16.0.0/12`-only allocation
rule, and `omnictl cluster template validate` passed either way.

**Related Issue:** #1086

## Root Cause

The base's `podSubnets` field was a key with only commented-out example values underneath it — i.e. an empty/null
list, not a missing key. Omni's template schema treats an empty `podSubnets` as "not set" and falls through to
Talos's machine-config default rather than rejecting the template, so `omnictl cluster template validate` gave no
signal that anything was wrong. The two existing per-cluster templates (`lungmen-akn`, `rhodes-akn`) both already
override `podSubnets` correctly, so this was a latent gap rather than a live defect — flagged during the #1038 code
review (Parse-Don't-Validate / Fail-Fast).

## Fix

Two changes, one per commit:

1. **[`catalog/omni/clustertemplates/base.yaml`](catalog/omni/clustertemplates/base.yaml)** — replaced the empty
   `podSubnets` field with a sentinel value (`MUST-OVERRIDE-SEE-IPAM.MD`) that is not a valid CIDR. A forgotten
   override now fails `omnictl cluster template validate` loudly on the cloned template instead of silently falling
   back to a Talos default. The header now documents that the base itself is not expected to pass `validate`
   (it never was applied directly, per its own "REFERENCE template — NOT applied directly" note).
2. **[`scripts/omni:template:lint`](scripts/omni:template:lint)** + **[`.github/workflows/push,pull_request.omni-template-lint.yaml`](.github/workflows/push,pull_request.omni-template-lint.yaml)**
   — the sentinel only catches a *completely forgotten* override. It doesn't catch a template that overrides
   `podSubnets`/`serviceSubnets`/`clusterDNS` with a value that's syntactically valid but still outside ADR-014's
   `172.16.0.0/12` range. The new script checks all three fields on every
   `projects/*/src/infrastructure/omni/*.clustertemplate.yaml` against that range (the reference base is excluded by
   path — it's never applied directly), and a new CI workflow runs it on any change under
   `projects/*/src/infrastructure/omni/`.

### Why not conftest/OPA

`catalog/opa/` policies are wired through `conftest --combine` against rendered `dist/` directories
(`.github/workflows/merge_group.compliance-full.yaml` / `push,pull_request.compliance-partial.yaml`), keyed on a
per-namespace directory matrix. Omni cluster templates are Talos DSL, not Kubernetes manifests, have no `dist/`
equivalent, and are single standalone files — forcing them through that dist-shaped matrix would be a much larger,
disproportionate change for what this issue needs. A small standalone script + dedicated workflow (same pattern as
`push,pull_request.vector-config-test.yaml`) is the proportionate fit.

### Behavioural Changes

- Cloning the base and forgetting to set `podSubnets` now fails `omnictl cluster template validate` instead of
  silently applying Talos's default CIDR.
- Any per-cluster template PR touching `projects/*/src/infrastructure/omni/` now runs an automatic ADR-014 CIDR
  check in CI.
- No change to the two existing per-cluster templates (`lungmen-akn`, `rhodes-akn`) — both already comply and pass
  the new lint (verified locally).

### Regression Risk

Low. `catalog/omni/clustertemplates/base.yaml` is a reference-only file, never applied directly — changing its
sentinel value cannot affect a live cluster. The new script and workflow are additive and were tested locally
against both real templates (pass) and a deliberately broken CIDR (correctly fails).

### Observability

`scripts/omni:template:lint` prints one `FAIL <file>: <field> ...` line per violation and a summary count; CI
surfaces the same output as a failed check.

## Testing Validation

- [x] `scripts/omni:template:lint` passes against the current `lungmen-akn` and `rhodes-akn` templates
- [x] `scripts/omni:template:lint` correctly fails against a deliberately-injected `10.244.0.0/16` podSubnets value
- [x] `trunk check --filter=-conftest` reports no issues on all changed files
- [ ] CI workflow run (will execute automatically once this PR is opened)

## Related Issues

Closes #1086

---

<sub>AI-assisted with claude-code:claude-sonnet-5 under human supervision</sub>

### CODE REVIEW

File: scripts/omni:template:lint [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+REPO_ROOT="$(git rev-parse --show-toplevel)"
+FAILURES=0
+
+# ADR-014: every Talos/Omni CIDR must live inside 172.16.0.0/12.
+in_adr014_range() {
+  local ip="${1%%/*}"
+  [[ ${ip} =~ ^([0-9]{1,3})\.([0-9]{1,3})\.[0-9]{1,3}\.[0-9]{1,3}$ ]] || return 1
+  local o1=${BASH_REMATCH[1]} o2=${BASH_REMATCH[2]}
+  [[ ${o1} == "172" ]] && ((o2 >= 16 && o2 <= 31))
+}
```
[the remark this answers is not in the document]
Comment (xunleii) [reply scripts/omni:template:lint]: Good catch — that's a real gap in the CI gate, not a nitpick. Added a valid_ipv4() helper validating all four octets (0-255) plus a 0-32 prefix-length check when a "/" is present, before applying the ADR-014 range check, in 15cf35060. Verified with your exact "172.16.999.0/19" example (now correctly caught) plus a fuller test matrix (invalid prefix, boundary values, valid in/out of range).

**Comment by xunleii** [comment]
Addressed the 3 suppressed findings from the second review pass in 585e2c06f:
1. Trailing `/` with an empty prefix (e.g. `172.16.0.0/`) now correctly fails instead of being treated as "no prefix given".
2. Leading-zero octets/prefixes (e.g. `172.016.0.1`) are now rejected by the regex, and all arithmetic is forced to base 10 (`10#$var`) so nothing can be misparsed as octal.
3. `check_field()` now captures and surfaces `yq`'s actual stderr on a query failure instead of silently reporting "missing or empty".

Verified with an extended test matrix plus real runs (clean pass on the actual templates, a re-injected `172.16.0.0/` regression test, and a malformed-YAML test confirming the real parse error now surfaces).

**Comment by xunleii** [comment]
Addressed the 2 suppressed findings from the third review pass in 03c136d46:
1. `REPO_ROOT` was assigned without checking `git rev-parse`'s exit status — outside a git checkout it would silently become empty and the script would report a false "all templates comply" with zero files actually checked. Now exits with an explicit error instead.
2. The lint accepted any in-range IPv4 for every field regardless of CIDR-vs-bare-IP semantics. `check_field()` now takes a "cidr"/"ip" kind and rejects a podSubnets/serviceSubnets value missing `/`, or a clusterDNS value carrying one.

Verified: clean run against real templates still passes, running from outside a git checkout now fails loudly, and both new injected cases (bare podSubnets IP, prefixed clusterDNS IP) are correctly caught.

**Comment by xunleii** [comment]
Addressed in acc5c9e53: the range check only validated the CIDR's base address, not that the whole block is contained within 172.16.0.0/12. Added a lower bound (prefix must be 12-32, not just <=32) since /12 is exactly ADR-014's supernet width. Verified with your `172.16.0.0/1` example (now rejected) plus boundary cases (/8, /11 rejected; /12 accepted; /19, /32 still pass).
